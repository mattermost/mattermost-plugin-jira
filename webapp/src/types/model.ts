export type ReactSelectOption = {
    label: string | React.ReactElement;
    value: string;
};

export type AllowedValue = {
    id: string;
    value?: string;
    name?: string;
};

export type FieldSchema = {
    type: string;
    system?: string;
    custom?: JiraFieldCustomTypeEnums;
    customId?: number;
    items?: string;
};

export type BaseField = {
    key?: string;
    name: string;
    required: boolean;
    schema: FieldSchema;
    allowedValues?: AllowedValue[];
};

export type SelectField = BaseField & {
    allowedValues: AllowedValue[];
};

export type StringArrayField = BaseField;

export type StringField = BaseField;

export type JiraField = SelectField | StringArrayField | StringField;

export type IssueTypeIdentifier = {id: string; name: string};

export type FilterField = {
    key: string;
    name: string;
    schema: FieldSchema;
    values?: ReactSelectOption[];
    userDefined?: boolean;
    issueTypes: IssueTypeIdentifier[];
};

export type IssueType = {
    id: string;
    name: string;
    fields: {[key: string]: JiraField};
    subtask: boolean;
}

export type JiraIssue = {
    id: string;
    key: string;
    name: string;
    fields: {[key: string]: JiraField};
}

export type Project = {
    key: string;
    issuetypes: IssueType[];
}

export type IssueTypeWithStatuses = {
    id: string;
    name: string;
    statuses: Status[];
}

export type IssueMetadata = {
    projects: Project[];
    issue_types: IssueTypeWithStatuses[];
}

export type Status = {
    id: string;
    name: string;
}

export type ProjectMetadata = {
    projects: ReactSelectOption[];
    issues_per_project: {[key: string]: ReactSelectOption[]};
    default_project_key?: string;
}

export enum JiraFieldTypeEnums {
    PROJECT = 'project',
    ISSUE_TYPE = 'issuetype',
    PRIORITY = 'priority',
    DESCRIPTION = 'description',
    SUMMARY = 'summary',
    LABELS = 'labels',
    ASSIGNEE = 'assignee',
    SECURITY = 'security',
    COMPONENTS = 'components',
    FIX_VERSIONS = 'fixVersions',
    AFFECTS_VERSIONS = 'versions',
}

export enum AvatarSize {
    XSMALL = '16x16',
    SMALL = '24x24',
    MEDIUM = '36x36',
    LARGE = '48x48',
}

export type JiraUser = {
    self: string;
    accountId: string;
    avatarUrls: {[size in AvatarSize]: string};
    displayName: string;
    active: boolean;
    timeZone: string;
}

export enum JiraFieldCustomTypeEnums {
    SPRINT = 'com.pyxis.greenhopper.jira:gh-sprint',
    EPIC_LINK = 'com.pyxis.greenhopper.jira:gh-epic-link',
    EPIC_NAME = 'com.pyxis.greenhopper.jira:gh-epic-label',
    RANK = 'com.pyxis.greenhopper.jira:gh-lexo-rank',

    CASCADING_SELECT = 'com.atlassian.jira.plugin.system.customfieldtypes:cascadingselect',
    TEXT_AREA = 'com.atlassian.jira.plugin.system.customfieldtypes:textarea',
    TEXT_FIELD = 'com.atlassian.jira.plugin.system.customfieldtypes:textfield',
    SELECT = 'com.atlassian.jira.plugin.system.customfieldtypes:select',
    PROJECT = 'com.atlassian.jira.plugin.system.customfieldtypes:project',

    MULTI_SELECT = 'com.atlassian.jira.plugin.system.customfieldtypes:multiselect',
    RADIO_BUTTONS = 'com.atlassian.jira.plugin.system.customfieldtypes:radiobuttons',
    MULTI_CHECKBOXES = 'com.atlassian.jira.plugin.system.customfieldtypes:multicheckboxes',
    URL_FIELD = 'com.atlassian.jira.plugin.system.customfieldtypes:url',

    LABELS = 'com.atlassian.jira.plugin.system.customfieldtypes:labels',
    USER_PICKER = 'com.atlassian.jira.plugin.system.customfieldtypes:userpicker',
}

export enum FilterFieldInclusion {
    INCLUDE_ANY = 'include_any',
    INCLUDE_ALL = 'include_all',
    EXCLUDE_ANY = 'exclude_any',
    EMPTY = 'empty',
}

export type FilterValue = {
    key: string;
    values: string[];
    inclusion: FilterFieldInclusion;
}

export type ChannelSubscriptionFilters = {
    projects: string[];
    events: string[];
    issue_types: string[];
    fields: FilterValue[];
};

export type ChannelSubscription = {
    id: string;
    channel_id: string;
    filters: ChannelSubscriptionFilters;
    name: string;
    instance_id: string;
}

export enum InstanceType {
    CLOUD = 'cloud',
    SERVER = 'server',
}
export type Instance = {
    alias?: string;
    instance_id: string;
    type: InstanceType;
}

export type GetConnectedResponse = {
    data: {
        can_connect: boolean;
        instances: Instance[];
        is_connected: boolean;
        user: {
            connected_instances: Instance[];
            default_instance_id?: string;
        };
    };
    error?: Error;
};

export type APIResponse<T> = {
    error?: Error;
    data: T;
};

export type CreateIssueRequest = {
    instance_id: string;
    required_fields_not_covered: string[][];
    post_id: string;
    current_team: string;
    channel_id: string;
    fields: {};
};

export type SearchIssueParams = {
    jql?: string;
    fields: string;
    q: string;
    instance_id: string;
};

export type AttachCommentRequest = {
    post_id: string;
    current_team: string;
    issueKey: string;
    instance_id: string;
};

export type AllProjectMetadata = {
    instance_id: string;
    metadata: ProjectMetadata;
}[];

export type CreateIssueFields = {
    description: string;
    summary?: string;
    project: {key: string};
    issuetype: {id: string};
} & {[key: string]: JiraField};
