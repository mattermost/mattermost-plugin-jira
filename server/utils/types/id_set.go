// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package types

import (
	"encoding/json"
)

type ID string

func (id ID) GetID() ID      { return id }
func (id ID) String() string { return string(id) }

type IDArray []ID

func (p IDArray) Len() int             { return len(p) }
func (p IDArray) GetAt(n int) Value    { return p[n] }
func (p IDArray) SetAt(n int, v Value) { p[n] = v.(ID) }
func (p *IDArray) Ref() interface{}    { return p }
func (p IDArray) InstanceOf() ValueArray {
	inst := make(IDArray, 0)
	return &inst
}
func (p *IDArray) Resize(n int) {
	*p = make(IDArray, n)
}

var IDArrayProto = &IDArray{}

type IDSet struct {
	ValueSet
}

func NewIDSet(vv ...ID) *IDSet {
	i := &IDSet{
		ValueSet: *NewValueSet(&IDArray{}),
	}
	for _, v := range vv {
		i.Set(v)
	}
	return i
}

func (i *IDSet) Set(v ID) {
	i.ValueSet.Set(v)
}

func (i *IDSet) MarshalJSON() ([]byte, error) {
	return json.Marshal(i.IDs())
}

func (i *IDSet) UnmarshalJSON(data []byte) error {
	ids := []ID{}
	err := json.Unmarshal(data, &ids)
	if err != nil {
		return err
	}

	n := NewIDSet(ids...)
	*i = *n
	return nil
}
