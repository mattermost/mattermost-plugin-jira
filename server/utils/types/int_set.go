// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package types

import (
	"encoding/json"
	"sort"
)

type IntValue struct {
	ID    ID
	Value int64
}

func NewIntValue(id ID, value int64) IntValue {
	return IntValue{
		ID:    id,
		Value: value,
	}
}

func (ii IntValue) GetID() ID { return ii.ID }

type intArrayProto []IntValue

func (p intArrayProto) Len() int             { return len(p) }
func (p intArrayProto) GetAt(n int) Value    { return p[n] }
func (p intArrayProto) SetAt(n int, v Value) { p[n] = v.(IntValue) }
func (p *intArrayProto) Ref() interface{}    { return p }
func (p intArrayProto) InstanceOf() ValueArray {
	inst := make(intArrayProto, 0)
	return &inst
}
func (p *intArrayProto) Resize(n int) {
	*p = make(intArrayProto, n)
}

type IntSet struct {
	ValueSet
}

func NewIntSet(vv ...IntValue) *IntSet {
	i := &IntSet{
		ValueSet: *NewValueSet(&intArrayProto{}),
	}
	for _, v := range vv {
		i.Set(v.ID, v.Value)
	}
	return i
}

func (set *IntSet) Set(id ID, v int64) {
	set.ValueSet.Set(NewIntValue(id, v))
}

func (set *IntSet) Get(id ID) int64 {
	v := set.ValueSet.Get(id)
	if v == nil {
		return 0
	}
	return v.(IntValue).Value
}

func (set *IntSet) MarshalJSON() ([]byte, error) {
	m := map[ID]int64{}
	for _, id := range set.ids {
		m[id] = set.Get(id)
	}
	return json.Marshal(m)
}

func (set *IntSet) UnmarshalJSON(data []byte) error {
	m := map[ID]int64{}
	err := json.Unmarshal(data, &m)
	if err != nil {
		return err
	}

	keys := []string{}
	for key := range m {
		keys = append(keys, key.String())
	}
	sort.Strings(keys)

	n := NewIntSet()
	*set = *n
	for _, key := range keys {
		id := ID(key)
		set.Set(id, m[id])
	}
	return nil
}

func (set *IntSet) TestAsMap() map[ID]int64 {
	out := map[ID]int64{}
	for _, id := range set.ids {
		out[id] = set.Get(id)
	}
	return out
}
