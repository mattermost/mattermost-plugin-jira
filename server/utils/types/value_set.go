// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package types

import (
	"encoding/json"
	"sort"
)

type Value interface {
	GetID() ID
}

type Setter interface {
	Set(Value)
}

type Getter interface {
	Get(Value)
}

type ValueArray interface {
	Len() int
	GetAt(int) Value
	SetAt(int, Value)
	InstanceOf() ValueArray
	Ref() interface{}
	Resize(int)
}

type ValueSet struct {
	proto ValueArray
	ids   []ID
	m     map[ID]Value
}

func NewValueSet(proto ValueArray, vv ...Value) *ValueSet {
	i := &ValueSet{
		proto: proto,
	}
	for _, v := range vv {
		i.Set(v)
	}
	return i
}

func (set *ValueSet) From(other *ValueSet) {
	set.proto = other.proto
	set.ids = append([]ID{}, other.ids...)
	set.m = map[ID]Value{}
	for id, v := range other.m {
		set.m[id] = v
	}
}

func (set *ValueSet) Contains(id ID) bool {
	if set.IsEmpty() {
		return false
	}
	_, ok := set.m[id]
	return ok
}

func (set *ValueSet) Delete(toDelete ID) {
	if !set.Contains(toDelete) {
		return
	}

	for n, key := range set.ids {
		if key == toDelete {
			updated := set.ids[:n]
			if n+1 < len(set.ids) {
				updated = append(updated, set.ids[n+1:]...)
			}
			set.ids = updated
		}
	}
	delete(set.m, toDelete)
}

func (set *ValueSet) Get(id ID) Value {
	if set.IsEmpty() {
		return nil
	}
	return set.m[id]
}

func (set *ValueSet) GetAt(n int) Value {
	if set.IsEmpty() {
		return nil
	}
	return set.m[set.ids[n]]
}

func (set *ValueSet) Len() int {
	if set.IsEmpty() {
		return 0
	}
	return len(set.ids)
}

func (set *ValueSet) IDs() []ID {
	if set.IsEmpty() {
		return []ID{}
	}
	n := make([]ID, len(set.ids))
	copy(n, set.ids)
	return n
}

func (set *ValueSet) Set(vv ...Value) {
	if set.ids == nil {
		set.ids = []ID{}
	}
	if set.m == nil {
		set.m = map[ID]Value{}
	}

	for _, v := range vv {
		id := v.GetID()
		if !set.Contains(id) {
			set.ids = append(set.ids, id)
		}
		set.m[id] = v
	}
}

func (set *ValueSet) SetAt(n int, v Value) {
	if set.ids == nil {
		set.ids = []ID{}
	}
	if set.m == nil {
		set.m = map[ID]Value{}
	}
	id := v.GetID()
	if !set.Contains(id) {
		set.ids = append(set.ids, id)
	}
	set.m[id] = v
}

func (set *ValueSet) AsArray(out ValueArray) {
	if set.IsEmpty() {
		out.Resize(0)
		return
	}
	out.Resize(len(set.ids))
	for n, key := range set.ids {
		out.SetAt(n, set.m[key])
	}
}

func (set *ValueSet) IsEmpty() bool {
	if set == nil {
		return true
	}
	return len(set.ids) == 0
}

func (set *ValueSet) TestAsArray(out ValueArray) {
	if set.IsEmpty() {
		out.Resize(0)
		return
	}
	out.Resize(len(set.ids))
	for n, key := range set.ids {
		out.SetAt(n, set.m[key])
	}
}

func (set *ValueSet) TestIDs() []string {
	if set.IsEmpty() {
		return nil
	}
	n := []string{}
	for _, id := range set.IDs() {
		n = append(n, string(id))
	}
	sort.Strings(n)
	return n
}

func (set *ValueSet) MarshalJSON() ([]byte, error) {
	if set.IsEmpty() {
		return []byte("[]"), nil
	}
	proto := set.proto.InstanceOf()
	proto.Resize(len(set.ids))
	for n, id := range set.ids {
		proto.SetAt(n, set.m[id])
	}
	return json.Marshal(proto)
}

func (set *ValueSet) UnmarshalJSON(data []byte) error {
	proto := set.proto.InstanceOf()
	err := json.Unmarshal(data, proto.Ref())
	if err != nil {
		return err
	}

	set.ids = []ID{}
	set.m = map[ID]Value{}
	for n := 0; n < proto.Len(); n++ {
		set.Set(proto.GetAt(n))
	}
	return nil
}
