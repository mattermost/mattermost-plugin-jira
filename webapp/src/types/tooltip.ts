import {JiraUser} from './model';

export type TicketDetails = {
    assigneeName: string;
    assigneeAvatar: string;
    labels: string[];
    description: string;
    summary: string;
    ticketId: string;
    jiraIcon: string;
    versions: string;
    statusKey: string;
    issueIcon: string;
}

export type TicketData = {
    key: string;
    fields: TicketDataFields;
}

export type AvatarUrls = {
    '48x48': string;
}

export type TicketDataFields = {
    assignee: JiraUser | null;
    labels: string[];
    description: string;
    summary: string;
    project: {avatarUrls: AvatarUrls};
    versions: string[];
    status: {name: string};
    issuetype: {iconUrl: string};
}

export type IssueAction = {
    type: string;
    data: TicketData;
}
