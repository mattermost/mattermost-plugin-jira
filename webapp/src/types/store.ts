// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {GlobalState as ReduxGlobalState} from 'mattermost-redux/types/store';

import type combinedReducers from '../reducers';

export type GlobalState = ReduxGlobalState & {
    'plugins-jira': PluginState
};

export type PluginState = ReturnType<typeof combinedReducers>

export const pluginStateKey = 'plugins-jira' as const;
