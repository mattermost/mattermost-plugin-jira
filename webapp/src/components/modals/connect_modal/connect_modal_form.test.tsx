// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import {InstanceType} from 'types/model';

import ConnectModalForm from './connect_modal_form';

describe('components/ConnectModalForm', () => {
    const baseActions = {
        closeModal: jest.fn().mockResolvedValue({}),
        redirectConnect: jest.fn().mockResolvedValue({}),
    };

    const baseProps = {
        ...baseActions,
        visible: true,
        theme: {},
        connectedInstances: [
            {
                instance_id: 'https://something.atlassian.net',
                type: InstanceType.CLOUD,
            },
        ],
        installedInstances: [
            {
                instance_id: 'https://something.atlassian.net',
                type: InstanceType.CLOUD,
            },
            {
                instance_id: 'http://localhost:8080',
                type: InstanceType.SERVER,
            },
        ],
    };

    test('should match snapshot', () => {
        const props = {...baseProps};
        const wrapper = shallow<ConnectModalForm>(
            <ConnectModalForm {...props}/>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('should redirect on submit', async () => {
        const closeModal = jest.fn().mockResolvedValue({});
        const redirectConnect = jest.fn().mockResolvedValue({});
        const props = {
            ...baseProps,
            closeModal,
            redirectConnect,
        };
        const wrapper = shallow<ConnectModalForm>(
            <ConnectModalForm {...props}/>,
        );

        wrapper.instance().handleInstanceChoice('', 'http://localhost:8080');
        expect(wrapper.state().selectedInstance).toEqual('http://localhost:8080');
        expect(wrapper.state().error).toEqual('');
    });

    test('should show error when user is already connected to instance', async () => {
        const props = {...baseProps};
        const wrapper = shallow<ConnectModalForm>(
            <ConnectModalForm {...props}/>,
        );

        wrapper.instance().handleInstanceChoice('', 'https://something.atlassian.net');
        expect(wrapper.state().selectedInstance).toEqual('https://something.atlassian.net');
        expect(wrapper.state().error).toEqual('You are already connected to this Jira instance.');
    });
});
