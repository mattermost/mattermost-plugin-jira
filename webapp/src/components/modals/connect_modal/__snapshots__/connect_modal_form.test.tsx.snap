// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ConnectModalForm should match snapshot 1`] = `
<form
  onSubmit={[Function]}
  role="form"
>
  <ModalBody
    style={
      Object {
        "backgroundColor": undefined,
        "color": undefined,
        "padding": "2em 2em 3em",
      }
    }
  >
    <ReactSelectSetting
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "https://something.atlassian.net",
            "value": "https://something.atlassian.net",
          },
          Object {
            "label": "http://localhost:8080",
            "value": "http://localhost:8080",
          },
        ]
      }
      theme={Object {}}
    />
  </ModalBody>
  <ModalFooter
    style={
      Object {
        "padding": "2rem 15px",
      }
    }
  >
    <FormButton
      btnClass="btn-link"
      defaultMessage="Cancel"
      disabled={false}
      extraClasses=""
      onClick={[Function]}
      savingMessage="Creating"
      type="button"
    />
    <FormButton
      btnClass="btn btn-primary"
      defaultMessage="Connect"
      disabled={true}
      extraClasses=""
      saving={false}
      savingMessage="Creating"
      type="submit"
    />
    <p
      className="error-text"
    />
  </ModalFooter>
</form>
`;
